### ISSUE TITLE
Localisation: Greek

### ISSUE THREAD

**Issue by andylamp** [OptiKey/OptiKey#221]
@JuliusSweetland Greek translation strings from transifex are done, briefly check them so I can start the implementation of the Greek custom keyboard layout. Hopefully we will be able to include this translation in the next release.

Do let me know if there is anything to note while doing that.

Regards.

PS: Sorry for the long wait but I was held up with various things.

**Comment by JuliusSweetland** [comment]
Looks good. I can't review Greek, but the translations appear complete (with the possible exception of "Android Dark" and "Android Light" which are in English?). Great work.

The tasks are as follows:

**1.Transifex translation of UI strings:** Complete

**2.Keyboard layout:**
![image](https://cloud.githubusercontent.com/assets/4722561/12917755/2eece7fe-cf33-11e5-8041-a3e66deaf080.png)

This is just for reference - the keyboard in OptiKey doesn't need to be exactly like this. For other languages the character positions have been the same as the standard keyboard, but commonly used accented characters have been added as they need their own keys in OptiKey, i.e. we do not overload keys with alternative accented characters.

Is this the correct list of accented characters using in Greek: άέήίϊΐόύϋΰώ

Which of these characters are used commonly and need to be on the main (Alpha) keyboard?

See here for reference: http://greek.typeit.org/

**3.Word list:** We need one.

**4.Anything else:** Is Greek written like latin based languages, i.e one key press per character, or are Greek characters more like symbol based languages where it takes multiple key presses to output a character? At the moment only 1 key press = 1 character is supported correctly.

**BRANCH:** New branch "Greek" available here: https://github.com/OptiKey/OptiKey/tree/Greek

**Comment by andylamp** [comment]
@JuliusSweetland thanks for the reply, I'll address each one of your questions:

1) Cool -- that means I am okay to proceed, that's good news.

2) The keyboard looks like it should, although here keyboards usually have both Latin and Greek characters in each key that's different from the Latin ones. You can view such example in the following picture:

![here](https://upload.wikimedia.org/wikipedia/commons/thumb/c/cf/KB_Greek.svg/420px-KB_Greek.svg.png).

Can you embed this functionality in OptiKey's keyboard, if not already present.

3) I've got one, well actually I found that this one [here](http://www.winedt.org/Dict/) that looks sufficient (The UNICODE small one) -- take a look and tell me if that's OK.

4) No, it's definitely written like Latin based languages, no fancy Japanese wizardry required here.

**Comment by JuliusSweetland** [comment]
@andylamp 
2.The Greek keyboard in OptiKey only needs characters for typing Greek. I guess you have latin characters on your keyboards so you can switch between typing in Greek and English (or other latin languages). In OptiKey this workflow is achieved by going to the MENU and then the LANGUAGE keyboard to switch from Greek to English, type something in English, and then switch back to Greek. I may make it faster to switch between languages in future, but you shouldn't need to include all Greek AND latin characters on the same keyboard, unless latin characters are used routinely to write Greek words?

3.Looks good, but it's huge! 575,000 entries. Is it possible to trim it down? English is only 75k. The largest dictionary we are flirting with at the moment is Dutch @ 300k. I'm concerned that larger dictionaries could introduce performance issues. It may, however, be necessary to have a dictionary of this size, but you tell me.

4.Excellent as I haven't figures out how to integrate an IME (input method editor) yet!

**Comment by andylamp** [comment]
2) Okay, I just pointed this out because that's how most keyboards that are localized for Greece ship; I understand your logic behind it. Also no, they are not used as often; besides the duality exists only on letters that are different than the ones in Latin as you can see from the picture.

3) I don't know what to trim, I'll try and look for a smaller one but Greek is really weird and rich language so I'll try my best.

4) ;)

I think I'll first create the keyboard layout and then try to profile how it performs with the dictionary so we can evaluate the performance. Additionally on that front, can't we just load the required languages dictionaries in memory? Wouldn't that make faster look-ups? Also memory is kinda cheap these days so spending a couple more Mb's for storing the dictionaries in-memory wouldn't be that much of a problem...

Finally we have to define what is acceptable performance... I have a pretty beasty machine so I cannot really test on "low-end" stuff -- as I have none to test with.

**Comment by JuliusSweetland** [comment]
@andylamp The dictionaries are loaded into memory, but auto complete and word matching (and other operations) occur frequently and are reasonably expensive. It might be fine, but I've always tried to keep dictionary size down where possible. See how it feels and I can test on a slow laptop if you're happy.

**Comment by andylamp** [comment]
@JuliusSweetland Okay sweet, I'll complete the keyboard as soon as I can; then we can decide about the dictionary size.

**Comment by JuliusSweetland** [comment]
Sounds good

**Comment by andylamp** [comment]
@JuliusSweetland I am almost finished but I am having a couple of annoying issues while testing the keyboard. First of all when for example changing the keyboard from English to say German after I quite OptiKey my keyboard and mouse are messed up. For example mouse click highlights all of the test from start to click or my caps lock functionality is reversed. This is fixed by logging out from my account or restarting my computer -- is this behavior expected?

PS: The translation guide is in dire need or a touch-up which I will attempt to do when done with the translation process.

**Comment by JuliusSweetland** [comment]
@andylamp I don't believe the language change within OptiKey (specifically) is the reason for your keyboard being messed up - instead I think it's more a general issue with OptiKey when the process is killed. If you have modifier keys pressed down (e.g. SHIFT) then when you quit OptiKey using the Menu -> Quit method the modifier keys are all released. If you kill the process (or stop debugging within Visual Studio) OptiKey doesn't have a chance to tidy up after itself and Windows can be left thinking SHIFT/ALT/CTRL/WIN, or some combination of these, are still down, which is why clicking the mouse behaves strangely; windows thinks you are SHIFT+CLICKING, which highlights text, or similar. Same sort of thing with your CAPS LOCK being reversed, as CAPS+SHIFT = lowercase. 

The easiest way to undo this is to manually press the modifers once - they will all be correct after that as Windows recognises each key being pressed down (which does nothing as Windows thinks the key is down anyway) and then up (which it listens to, releasing the key).

RE: the translation guide - I agree. This was kindly added by someone else, but it could do with a refresh. Let me know if you will/won't have time and we can get this sorted out.

**Comment by andylamp** [comment]
@JuliusSweetland I'll definitely do it after I am finished with the process of translating so that I have a thorough understanding of what needs to be done. I'll keep you posted, hopefully I will be done within the day.

**Comment by andylamp** [comment]
@JuliusSweetland when changing to the Greek Keyboard I see a massive slowdown on the interface responsiveness and GC is going crazy for no apparent reason (like 5 secs to respond vs instant on the other langujages). I have trimmed the dictionary to 3 entries to eliminate that issue, do you have any idea what might be causing this behavior?

I have profiled the application and the method that's taking the most time (even with the trimmed dic) is the one that loads the dictionary as you can see from here:

![perl_call](https://cloud.githubusercontent.com/assets/2177249/12949450/8b0d2b88-d010-11e5-9f0f-0da40d4bde1e.JPG)

I have not looked into the function's implementation but before I dig into it could you please take a look using the dictionary I provided (trim it to like 10 entries and same thing happens) -- as I am quite sure the problem lies there.

Let me know once you look into it.

**Comment by JuliusSweetland** [comment]
@andylamp I'm not able to debug right now, but when the Greek dictionary was first loaded it would have created a copy in your roaming directory, which would be re-loaded each time, so trimming the source dictionary would not fix the problem. I'm guessing, but I think the delay is loading the huge dictionary each time. Try trimming the source dictionary GreekGreece.dic to 10 entries and removing your user dictionaries (see here for a guide: https://github.com/OptiKey/OptiKey/wiki/Factory-reset-the-dictionary).

Let me know if that helps. If it does then I think we've found an upper limit on dictionary size!

**Comment by andylamp** [comment]
@JuliusSweetland would that be located in /bin dir? I've deleted these directories altogether and I've put the always copy option on the dictionary resource file... so this should not be the case, unless I am mistaken... 

Let me know

**Comment by JuliusSweetland** [comment]
@andylamp The source dictionary is read from the bin directory, but only the first time you use a language. As part of loading the dictionary into memory the source dictionary file is copied (and transformed slightly) into a new file while roams with your user profile. This new copy is user specific and so can be edited (in the Management Console), it keeps track of your word usage counts, and is where any new words and phrases that you add are stored. This is all done on a user specific copy so that other users on the same computer will not be affected. The file can be found here:

C:\Users\YOUR_USER_NAME\AppData\Roaming\JuliusSweetland\JuliusSweetland\OptiKey\Dictionaries

You need to delete it and then the source dictionary will be loaded again on next run of OptiKey.

**Comment by andylamp** [comment]
@JuliusSweetland I was not aware of that fact, it helped -- that's the issue it seems. Thanks for the heads up. I'll definitely will have to trim the dictionary quite a bit it seems.

On a side note can we create a script so that we can remove all traces with the push of a button? I mean I thought clean solution would do the trick and spent quite sometime trying to figure out what was wrong.

**Comment by JuliusSweetland** [comment]
@andylamp Good news. It might be possible to optimise the parsing of the dictionary, but I think the easiest approach would be to limit the dictionary first.

I don't have a script, but I do sometimes keep an explorer window open and delete the contents of this folder:

C:\Users\YOUR_USER_NAME\AppData\Roaming\JuliusSweetland

That's effectively a factory restore each time. It removes custom dictionaries, changed settings, and log files.

**Comment by JuliusSweetland** [comment]
@andylamp Excellent PR (https://github.com/OptiKey/OptiKey/pull/222). My only comments are:

1.The language keyboard layout is incomplete (I think you know this) as it spills over into the 3rd row, which doesn't exist. I have today merged the Dutch branch into master, so if you update from master the Language keyboard has been changed anyway to a 3x3 grid. You might want to change it to 3x4 (3 rows, 4 columns) and slot in Greek.

2.The set of multi-key selection characters looks good (ασδφγηξκλ;ςερτυθιοπζχψωβνμ)

Really impressed though. Apart from needing a smaller dictionary this looks to be almost there. 

How does it feel when you test it?

**Comment by andylamp** [comment]
@JuliusSweetland 
1. Yeah, it was just for testing -- placement is easily fixed but before I do that since I want to follow alphabetical sorting shall I snap-it just after German and shift the others one place down (as I've done already, so Russian would be shifted to the third row).

2. Cool, I am glad it's okay

As I said in my previous message it seems to me that it works fine... but fingers crossed ;).

**Comment by JuliusSweetland** [comment]
That sounds good re language keyboard layout, but update from master first
(or let me know if you can't and I'll do it for you).

**Comment by andylamp** [comment]
@JuliusSweetland can you do that? then I'll try to merge with upstream on my local repo.

**Comment by JuliusSweetland** [comment]
Sure

**Comment by andylamp** [comment]
@JuliusSweetland cool, I'll try to get it sorted out by today and experiment to find a suitable dictionary size by profiling. Hopefully the translation will be finished by today so you can merge with master -- then I'll get on with the task of overhauling the translation guide.

**Comment by andylamp** [comment]
@JuliusSweetland let me know when you merge with `master` branch so that I can start pulling the changes locally.

**Comment by JuliusSweetland** [comment]
@andylamp Merged. I haven't changed Languages.xaml so the grid needs to be extended by an extra row and a Greek key added in the right place.

**Comment by JuliusSweetland** [comment]
@andylamp FYI There's an extra step to localisation that should be covered in the Localisation.md document: 

The setting 'KeySelectionTriggerFixationCompleteTimesByKeyValues' needs a new item added to it. This can be done in app.config. If you then reload the project in Visual Studio it should prompt you to update Settings.settings from the app.config change, otherwise you need to manually update the same setting. Here's an example of the addition that needs to be made. The functionKey value is the enum FunctionKeys value as a string, while the rest should be the same as below:

```
            <item>
              <key>
                <keyValue>
                  <functionKey>DutchNetherlands</functionKey>
                </keyValue>
              </key>
              <value>
                <ticks>PT1.75S</ticks>
              </value>
            </item>
```

**Comment by andylamp** [comment]
@JuliusSweetland my version of `App.config` does not have an entry named `KeySelectionTriggerFixationCompleteTimesByKeyValues`, can you please elaborate a bit more?

Also `Settings.settings` has a really weird format for this entry, it uses the HTML codes for special characters; for example `<dictionary>` is typed as: `&lt;dictionary&gt;` -- is this intentional?

Let me know.

**Comment by JuliusSweetland** [comment]
If it's not in your app.config don't worry, just edit it in the
settings.settings file. The xml friendly characters e.g. &lt; seem to have
been converted from my angle brackets automatically. If you copy/paste the
setting to notepad it's easier to edit and paste back into visual studio. I
am happy to edit the setting for you tomorrow if you're not sure.

**Comment by andylamp** [comment]
@JuliusSweetland oh okay, I mean I was wondering about that since the rest of the file is fine -- if this part is changed to normal brackets (as they are considerably more readable) will this be a problem? I'll edit the `Settings.settings` file but wanted the clarification. It's not that I am not sure but it's a big project and I want to be sure of the changes I make -- better be safe than sorry.

**Comment by JuliusSweetland** [comment]
To be honest I'm not sure. Give it a go - it should blow up fast if it's
wrong.

**Comment by andylamp** [comment]
@JuliusSweetland Okay, I'll give it a shoot tomorrow (although I think it will probably be fine since the rest of the file uses brackets normally). Hopefully, we will be done tomorrow with the technicalities and be left only with the dictionary size issue.

**Comment by JuliusSweetland** [comment]
@andylamp I have merged from master again, updated the Languages keyboard to include all languages (there are 9 now). Everything looks good to me so yes, the outstanding task would be the Greek dictionary.

I quickly tested typing auto-complete and multi-key selection and both appear to work, but can you double check with a bigger dictionary that everything "feels" right and works well with Greek?

I also notices two keys that looks odd - can you comment on these (shown here in upper and lower case):

![image](https://cloud.githubusercontent.com/assets/4722561/13027684/a94bac2c-d250-11e5-8c73-9663d8cd0cd7.png)

![image](https://cloud.githubusercontent.com/assets/4722561/13027685/b05d89c2-d250-11e5-8419-f90e9c3fa107.png)

**Comment by andylamp** [comment]
@JuliusSweetland Oh, yep the bottom is the correct one, the top has the keys when you have pressed the `Shift`-modifier key. Probably it was a mistake on my end.

**Comment by JuliusSweetland** [comment]
@andylamp are you happy to correct as I'm not familiar with Greek letters?

**Comment by andylamp** [comment]
@JuliusSweetland well, my initial commit was correct -- basically, the two pictures you posted above are correct. One is for `Shift`-pressed (top) and one for `Shift`-released (bottom). You can see that [here](http://gate2home.com/Greek-Keyboard) or even in the built in Greek windows keyboard.

**Comment by JuliusSweetland** [comment]
Ok understood. Are both keys letters, or is the leftmost key a
colon/semicolon? If so I would suggest that we remove that key, unless a
colon/semicolon is used often in Greek?

**Comment by andylamp** [comment]
@JuliusSweetland it's not used more or less than it's used in other Latin based languages, say for example English -- but it's there for some reason (it was not my decision to make ;)). The equivalent to `q` in Greek has only characters, `;` for `Shift`-up and `:` for `Shift`-down. The equivalent to `w` is needed as you get the commonly used terminal `ς` and `΅` which is used sparsely in some words. 

Hope this helps.

**Comment by JuliusSweetland** [comment]
So I could remove the :; key without impacting the ability to type Greek,
but the other key is used sometimes and should remain? If that's right I'll
make the change soon.

**Comment by andylamp** [comment]
@JuliusSweetland yes, you are correct. I don't think it would impact the ability to type Greek -- but one key that's missing which is essential to write Greek is the equivalent of `;` (the one that's located just after `l`) in the English keyboard as you need that letter for punctuation.

**Comment by JuliusSweetland** [comment]
You mean this: ΄ (a sort of tick character, which is different from an
apostrophe ')?

**Comment by andylamp** [comment]
@JuliusSweetland no, I mean this key:

![key_p](https://cloud.githubusercontent.com/assets/2177249/13028352/64deb4ba-d275-11e5-8b0b-3fa3777687d1.jpg)

**Comment by JuliusSweetland** [comment]
@andylamp I'm struggling to find that character. I tend to refer to this site: http://greek.typeit.org/ and it doesn't look like that character exists on there?

Also can I double check that you've added all diacritic characters that are on that site?

**Comment by andylamp** [comment]
@JuliusSweetland that character does not exist on that site because it has separate keys for punctuated letters (they are the above row where each of the letters has `'` over it) -- this is not normal, and most keyboards don't have them. That character as I said is the one that's mapped to the circled key indicated in the English keyboard I pasted in my previous post. Take a look that the keyboard presented in this site I mentioned earlier [here](http://gate2home.com/Greek-Keyboard), there it has the character you seek. FInally I attach here a screenshot of the character in the on-screen windows keyboard:

![key_p1](https://cloud.githubusercontent.com/assets/2177249/13029560/257d36fc-d298-11e5-9f2f-9a96a4204595.jpg)

**Comment by JuliusSweetland** [comment]
Oh I see, so it's a decorator key, i.e. you press it and then a letter and
it accents that letter. We don't follow that workflow in OptiKey. No key
has alternate characters on it that you press with the aid of
ctrl/shift/alt, or three concept of a decorator. Instead each common
accented character has it's own key, as per the typeit link I posted. Do
you think that's possible with Greek?

What I have done so far is include all the common accented letters on the
main alpha keyboards, and then include all possible accented characters for
the current language at the beginning of the didactic keyboard. You'll see
what I mean if you switch to French.

**Comment by andylamp** [comment]
@JuliusSweetland it's possible yes, but it would not be a normal keyboard and it requires a significant amount of additional keys. I'll check French and I'll let you know.

**Comment by JuliusSweetland** [comment]
Great, thanks

**Comment by JuliusSweetland** [comment]
@andylamp I haven't made any changes on the Greek branch, but I've been thinking about the problem and have a few thoughts/questions:

1.What's the difference between the Greek keyboard layout in OptiKey and in your screenshot:
![key_p1](https://cloud.githubusercontent.com/assets/2177249/13029560/257d36fc-d298-11e5-9f2f-9a96a4204595.jpg)

2.Is the second circled character a letter or another "decorator" that you press, followed by another letter? I ask because the greek.typeit.org keyboard has letters that appear to use that character: ΐΰ (lower) and ΐΰ (upper)?
![image](https://cloud.githubusercontent.com/assets/4722561/13027684/a94bac2c-d250-11e5-8c73-9663d8cd0cd7.png)
If it is a decorator then does it share the same key as ς because the lower and uppercase versions of ς are the same?

Speaking generally with the accented letters (diacritical characters) we have two options; we can change the way accented characters work across the whole of OptiKey and introduce "decorator" keys (which I am keen to avoid at the moment unless we have to do this), OR, we lose the decorator keys and introduce all accented letters with their own keys. As Greek appears to have lots of them we can make this more tolerable by only including the commonly used letters on the alpha keyboards and then add all accented letters as the first letters on the diacritic1 keyboard (as per French, German, Dutch).

Correct me where I go wrong, but these are the accented characters (all lowercase): άέήίϊΐόύϋΰώ which is 11 characters. This is a lot. Can be isolate a set of commonly used characters out of this set?

**Comment by andylamp** [comment]
@JuliusSweetland I am not really keen on lettings characters out as *every* word in Greek has *at least* one decorator which is used for punctuation. The way I see it going is either you introduce decorator keys or leave punctuation out altogether. The latter can lead to major issues, so I don't really recommend it, as most text-to-speech programs use this information to correctly pronounce the word.

**Comment by JuliusSweetland** [comment]
@andylamp Ok - I guess I'm going to be introducing decorator keys then. I'll look into the problem and get back to you.

Regarding you saying the difference between the screenshots is that shift is up/down - we should only be including the upper and lower versions of the **same** character on a key as the ShiftUpText and ShiftDownText properties, i.e. you shouldn't change the meaning of a key depending on whether the shift key is up/down.

**Comment by andylamp** [comment]
@JuliusSweetland I can help you with that. I am quite keen with C# and I was planning anyways to help a bit more than just translating...

Well, this might be a problem as this duality does happen in Greek keyboards ;) -- I am always talking about the **standard** ISO Greek keyboard layout -- as the one you previously presented [here](http://greek.typeit.org/) certainly isn't ;) .

**Comment by JuliusSweetland** [comment]
@andylamp Great. Let's both look into the problem and discuss it. I'm thinking of creating a set of acceptable accented characters per language, along the lines of the MultiKeySelectionKeys arrays in KeyValues. Then add the decorator keys to the keyboards, which would behave like modifier keys, in that they can be UP or DOWN (KeyDownState). Only one can be down at a time (I think this rule will hold) and the down decorator key would then be attached to the next letter that is typed, if that letter is in the set of acceptable accented characters for the current language, OR release the decorator key back to UP if anything else is input (including clear, backone, backmany, etc). The decorator variable would also have to be cleared if ANY other key was pressed. The "decorator key down" state would also have to be checked by the input service to prevent new multi-key selections.

Those are my initial thoughts.

**Comment by andylamp** [comment]
@JuliusSweetland I'll think about it more carefully because we need to take in-account all of the flavors that possible languages will have -- I can only speak for those languages that I am somewhat familiar with, these are: English, Greek, German and Japanese. In most languages decorator keys go over vowels of a digraph of those i.e. `ά` or `αί` in Greek. Generally we must have the ability to filter out the possible combination of letters that will accept said vowels and "gray-out" the rest of the alphabet for faster and more concise typing experience. The rule that one decorator will be active at any given time sounds valid -- but I am not an expert in linguistics so I wouldn't know for sure if we might have to alter it in the future (hopefully, we won't...). Finally, using decorator as a modifier key would be the way to go as it's in line with most implementations out there -- so it's a tested and trusted method.

**Comment by JuliusSweetland** [comment]
@andylamp We can use the KeyEnabledStates class to grey out keys when a "Diacritical mark" (decorator) key is down, including other decorator keys. This will limit the user to pressing one decorator at a time, followed by the acceptable letter for the current language. All of that logic can be achieved in KeyEnabledStates.

The KeyboardOutputService can then handle the addition of the mark to the letter (or ignore the decorator if something other than a letter is selected), followed by releasing all decorator keys, in the same way that the modifier keys are automatically released. It actually shouldn't be a big change.

The wikipedia article on diacritic marks is useful: https://en.wikipedia.org/wiki/Diacritic

**Comment by JuliusSweetland** [comment]
Conveniently that wiki page has a bar on the right listing diacritic marks. One section is "Latin and Greek":

accent
acute( ´ )
double acute( ˝ )
grave( ` )
double grave(  ̏ )
breve( ˘ )
inverted breve(  ̑ )
caron, háček( ˇ )
cedilla( ¸ )
circumflex( ˆ )
diaeresis, umlaut( ¨ )
dot( · )
hook, hook above(   ̡   ̢  ̉ )
horn(  ̛ )
iota subscript(  ͅ  )
macron( ¯ )
ogonek, nosinė( ˛ )
perispomene(  ͂  )
ring( ˚, ˳ )
rough breathing( ῾ )
smooth breathing( ᾿ )

**Comment by andylamp** [comment]
@JuliusSweetland sounds right, we can built a beta to test it out in a couple of days since browsing through the code it should not be a huge change. These are okay for Latin based languages -- I have absolutely no idea how to handle these issues when we talk about weird and quirky languages like Arabic or the likes.

**Comment by JuliusSweetland** [comment]
@andylamp We can tackle the complexities of Arabic, Chinese, etc at a later date. That's going to require full IME support I would guess. For now I've got a reasonably solid plan for how to display diacritic marks on their own keys (we need to use different unicode characters for the Text and for the Value properties, as unicode has one set of characters for display and another for combining with other characters. If you want to experiment this is a LinqPad test program which shows how composition works:

```
void Main()
{
       Compose("\x0061\x030C");
       Console.WriteLine("------");
       Compose("a ");
       Console.WriteLine("------");
       Compose("\x0061\x0341");
       Console.WriteLine("------");
       Compose("a ");
}

private void Compose(string decomposed)
{
       Console.WriteLine("Decomposed:{0}", decomposed);
       foreach (char c in decomposed)
           Console.WriteLine("\t{0} => U+{1:X4}", c, (int)c);
       string composed = decomposed.Normalize(NormalizationForm.FormKC);
       Console.WriteLine("Composed:{0}", composed);
       foreach (char c in composed)
           Console.WriteLine("\t{0} => U+{1:X4}", c, (int)c);
}
```

Output:

```
Decomposed:ǎ
  a => U+0061
  ̌ => U+030C
Composed:ǎ
  ǎ => U+01CE
------
Decomposed:ǎ
  a => U+0061
  ̌ => U+030C
Composed:ǎ
  ǎ => U+01CE
------
Decomposed:á
  a => U+0061
  ́ => U+0341
Composed:á
  á => U+00E1
------
Decomposed:á
  a => U+0061
  ́ => U+0341
Composed:á
  á => U+00E1
```

(Some of the combination characters are not displaying correctly, but hopefully you can see what I am doing here)

**Comment by JuliusSweetland** [comment]
This is also useful (for me): https://en.wikipedia.org/wiki/Greek_diacritics

**Comment by andylamp** [comment]
@JuliusSweetland yep, I advised that page myself -- but it's really to detailed as modern Greek mostly uses two characters for punctuation `΄` and `΅` also most words have only *one* punctuation point. We can use this to simplify our implementation quite a bit.

**Comment by JuliusSweetland** [comment]
@andylamp I've done a bit of experimentation and have a few observations/questions:

* KeyEnabledStates - diacritics are case sensitive (e.g. Y can take a diaeresis only when uppercase)
* Letter + diaeresis + accent does not compose into a single character - it is still 2 characters (a letter with diaeresis + an accent). 
* It is order specific, i.e. letter + acute accent + diaeresis doesn't compose correctly. Only combine ONE diacritic with ONE letter.

```
	a + diaeresis + acute accent:
	Composed:ä́
	  ä => U+00E4
	  ́ => U+0301
	---
	a with diaeresis + acute accent:
	Composed:ä́
	  ä => U+00E4
	  ́ => U+0301
	---
	a + acute accent + diaeresis:
	Composed:á̈
	  á => U+00E1
	  ̈ => U+0308
	---
	a with acute accent + diaeresis:
	Composed:á̈
	  á => U+00E1
	  ̈ => U+0308
```

* Two possibilities for letter/diacritic input order:
	a)Type the combining mark first - manually compose into a single character before publishing etc
	b)Type the combining mark second - this will work automatically, but requires the enabled state of the diacritic marks to be based on the last character being a letter with NO diacritic marks (as they do not combine correctly). There are then two options:
		i.Leave the letter and combining characters as seperate characters - this will require changes to some string methods/operations, e.g. need to modify the back one/back many logic to backspace through combining characters.
		ii.Manually compose into a single character before publishing etc

* The Unicode standard for diacritic marks: http://unicode.org/charts/PDF/U0300.pdf

**Comment by JuliusSweetland** [comment]
Here's the LinqPad test code:
```
void Main()
{
	Console.WriteLine("a + diaeresis + acute accent:");
	Compose("\x0061\x0308\x0301", false); //a + diaeresis + accent
	Console.WriteLine("---");
	Console.WriteLine("a with diaeresis + acute accent:");
	Compose("\x00E4\x0301", false); //a with diaeresis + accent
	Console.WriteLine("---");
	Console.WriteLine("a + acute accent + diaeresis:");
	Compose("\x0061\x0301\x0308", false); //a + accent + diaeresis
	Console.WriteLine("---");
	Console.WriteLine("a with acute accent + diaeresis:");
	Compose("\x00E1\x0308", false); //a with accent + diaeresis
	Console.WriteLine("---");
	Console.WriteLine("a + acute accent with diaeresis:");
	Compose("\x0061\x0344", false); //a + accent with diaeresis
}

private void Compose(string decomposed, bool outputDecomposedString)
{
	if(outputDecomposedString)
	{
		Console.WriteLine("Decomposed:{0}", decomposed);
    	foreach (char c in decomposed)
        	Console.WriteLine("\t{0} => U+{1:X4}", c, (int)c);
	}
    string composed = decomposed.Normalize(NormalizationForm.FormKC);
    Console.WriteLine("Composed:{0}", composed);
    foreach (char c in composed)
    	Console.WriteLine("\t{0} => U+{1:X4}", c, (int)c);
}
```

**Comment by JuliusSweetland** [comment]
Using the above discoveries this is what I'm thinking:

1. Dead keys (diacritic keys which are combined with a subsequent letter to produce a letter+diacritical mark) are not handled correctly by VkKeyScanEx (as mentioned in http://www.siao2.com/2006/09/10/748742.aspx), which is what I use to attempt to simulate "real" key strokes. Instead these combining characters will end up being sent as text, which may then be correctly combined with the previous output letter, but this is dependent on the target application. Instead I think letters and diacritics must be combined BEFORE we attempt to publish them.

2. Ordering - both 'letter before diacritic', and 'diacritic before letter' could be achieved if we are going to manually concatenate the letter and diacritic, and then compose them into a single letter, but I think the UX would be better if we follow the way physical dead keys work; you press the dead key and then the letter. I think this is nicer as people might already be familiar with this, and also we can disable all keys except the letters which can accept that diacritic, guiding people towards the correct choices.

3. As I already said, we should have 3 diacritic keys (acute accent, diaeresis and acute accent with diaeresis) and limit the user to one diacritic mark with one letter to avoid ordering problems where users attempt to combine diacritics in an unsupported ordering.

**Comment by andylamp** [comment]
@JuliusSweetland I agree on these completely -- will form a more complete response after I get back from work ;).

**Comment by JuliusSweetland** [comment]
@andylamp I also wanted to ask you; I've learned that the sigma character (Σ) has 2 lowercase equivalents; ς and σ - this is a new concept to OptiKey. I think there are problems using the ς key for an unrelated symbol when shift is down (you added the acute accent + diaeresis) as the shift down character is not visible on the keyboard when shift is up. On the other hand adding sigma to both keys results in 2 sigma keys when shift is down. Do you have any idea what we could do?

**Comment by JuliusSweetland** [comment]
@andylamp One more thing:

The diacritic characters that we have to use are called "Combining" characters, but these cannot be displayed on their own as they have to combine with another character (a spacing character) before they are rendered. Most combining diacritic characters have another non-combining character which we can use as the Text property of the keyboard Key, however, not all have an equivalent. So... it turns out you can combine a combining character with a space and it will render, which makes perfect sense. For example, rather than using the acute accent character (XML escaped Unicode value &#x00B4;), you can instead use the same combining character that we will use in the Value property like this; &#x0020;&#x0301; which is a space followed by the combining acute accent.

The key would therefore look a bit like this (the Value would :

```
<controls:Key Grid.Row="2" Grid.Column="1" Grid.ColumnSpan="2"
              Text="&#x0020;&#x0301;" 
              SharedSizeGroup="KeyWithSingleLetter"
              Value="{x:Static models:KeyValues.AcuteAccentKey} />
```

**Comment by JuliusSweetland** [comment]
@andylamp I've added all latin and greek diacritic marks from the Wikipedia article to KeyValues, as well as a new char extension method to determine if the captured string contains a combining character and should be composed. For use in the KeyboardOutputService to compose letter + diacritic before publishing.

**Comment by andylamp** [comment]
@JuliusSweetland Okay let's tackle some things (that's only for `Greek` though).

* We do not use diacritics on capital letters; most have them by convention but are not used -- syntactically they are correct but we don't use them.

 * `σ` and `ς` are the same letter albeit serve different purposes -- the first is used to express "sigma" inside a word while the latter is used for "sigma" as an ending to the word.

 * we should only accept one diacritic above a letter and (at least for Greek) at most two for each word.

 * The less diacritic and complexities we add the better -- three are more than enough for Greek (not aware of other languages).

I've done some experiments on my end and Greek are sufficient with just two diacritics (we are omit somethings -- they are mostly remnants of the Ancient Greek and are not used, although textbooks have them for completeness). So we can get rid of some cases and use either a dieresis or accent diacritic but not both.

The changes you made are in upstream/master? Can you merge?

**Comment by JuliusSweetland** [comment]
@andylamp Thanks - I've got a solution in mind and will try to get something functional tomorrow.

The sigma lower/upper ambiguity might need some more discussion as OptiKey doesn't support a key having more than one value. At the moment (and I know this will look weird) I plan to have two sigma keys, with different lowercase characters.

```
Console.WriteLine("σ".ToUpper()); //Prints "Σ"
Console.WriteLine("ς".ToUpper()); //Prints "Σ"
Console.WriteLine("Σ".ToLower()); //Prints "σ"
```

**Comment by andylamp** [comment]
@JuliusSweetland Okay, let me know how you plan on doing it based on OptiKey's existing limitations and I'll give you my thoughts.

**Comment by JuliusSweetland** [comment]
@andylamp Changed made and committed. There are two diacritic keys which act as dead keys (on the top row of the keyboard on the far left and right, but please note that the far right diaeresis is not displaying correctly at the moment) - you press one and then a letter to combine them and output the combined character. You can also combine the accent and diaeresis if shift is not down (as it only combines on lower case letters).

Note that there are 2 sigmas - let me know what you think about this.

Outstanding changes:
* Diaeresis text is not displaying correctly on Alpha Keyboard
* ConversationAlpha has not been updated - we can do this once Alpha is complete
* Greek customised diacritic keyboards are required

**Comment by andylamp** [comment]
@JuliusSweetland Okay, got the commit. Will review code changes and see if I have anything else to add. I'll also test it a bit to see if it's working as intended.

Also can you elaborate on why the dieresis text isn't displayed as intended?

The last outstanding issue is easy to fix -- let's tackle the first two initially.

**Comment by JuliusSweetland** [comment]
@andylamp Great stuff. I've solved the lack of text on the diaeresis key - it was because the Roboto font cannot correctly render all of the Unicode diacritic marks. I have just added a new commit which adds the concept of a compatibility font (Charis SIL) which can render things the user chosen font may not be able to. I turn it on only for keys which need it. I think it works well. It is also a good looking font, so I've made it available for users to select in the Management Console.

I need to repair the tests that I've broken and then we have the issue of the double sigma keys.

Lastly, we still need a reasonably sized Greek dictionary and then we can do some final testing.

**Comment by andylamp** [comment]
@JuliusSweetland Great! Cool, I'll take a look at the whole commit first thing tomorrow -- I'd agree that having more than one keyboard *when chiefly talking about disabled people* for languages that commonly use diacritics can be unproductive -- especially in `Greek` and other similar use-cases.

**Comment by JuliusSweetland** [comment]
@andylamp Have you had a chance to review?

The outstanding issues are:

1. We have 2 x sigma keys - is this a problem, and if so do you have an idea
2. We need a dictionary word list
3. The conversation alpha keyboard needs to be updated to include the dead keys

I'll then migrate all other languages over to have dead keys for diacritics (where appropriate) and release the whole lot together.

**Comment by andylamp** [comment]
@JuliusSweetland unfortunately no :( -- I was really swamped today, will get back to it as soon as possible and I'll let you know pronto.

**Comment by JuliusSweetland** [comment]
Thanks. I'm eager to hear your feedback on the new diacritic dead keys.

**Comment by andylamp** [comment]
@JuliusSweetland I've started testing the new commit yesterday; found some issues will update when I complete my testing.

**Comment by JuliusSweetland** [comment]
@andylamp How is the testing going? I may have some time later today to work on the issues so let me know if you have some feedback already.

**Comment by andylamp** [comment]
@JuliusSweetland I've tested it -- but I cannot get the application to display the punctuation on the words correctly. Fiddled with it. no luck. Also I should note that while punctuation for capital letters not widely used for *all* capital words, it *is* used with normal words. So for example `When` in Greek `'Όταν` should have a dieresis initially which again I could not perform.

Also the Greek language speaking is broken, is this a bug or I should install a specific Windows feature/package for it to work?

Let me know! (and sorry for the delay)

**Comment by JuliusSweetland** [comment]
The steps to test should be:

1.Open OptiKey and notepad. Notepad should be the focused application so we
can type into it with OptiKey.

2.On the main alpha keyboard select a diacritic key. When you select it it
will be pressed down and turn blue. Nothing will be output at this point. A
number of keys will be disabled, leaving only letters which can be
decorated with that diacritic.

3.Select one of the enabled letter keys. A letter with diacritic should be
output into OptiKey's scratchpad and typed into notepad.

Let me know what fails for you.

In terms of the Greek TTS language - you need to have a Greek voice
installed in Windows. This can probably be installed as part of the Windows
language pack. You then need to select it in the OptiKey management studio
under "Sounds". Otherwise if you leave OptiKey using an English voice you
probably won't hear anything when you select Speak as the English TTS voice
won't be setup to pronounce Greek words.

**Comment by JuliusSweetland** [comment]
@andylamp Is Greek working for you now?

**Comment by JuliusSweetland** [comment]
@andylamp Hi Andy - I've got some changes that I want to make to the other languages in the Greek branch, after which I'd like to start testing the branch for release. Can you let me know how you are getting on with sourcing a decent Greek dictionary and your testing in general? Thanks.

**Comment by andylamp** [comment]
@JuliusSweetland I'll try to give some feedback today; I have been caught up by other matters and didn't have enough time to do so. I'll get back to you as soon as possible.

PS The dictionary issue is not really simple, as most dictionaries that I searched are way too large. I'll see what I can do to amend it -- but I am not really sure on what to leave out and what to keep.

**Comment by JuliusSweetland** [comment]
@andylamp Great, thanks. If we can use a stripped down dictionary for testing and get the general functionality approved, then I will help you find a dictionary.

**Comment by andylamp** [comment]
@JuliusSweetland I have some good news; the issues were superficial, I was a couple of remote commits behind on my local repo and didn't get the latest changes (was confused by having no version bump -- so I thought I was OK, but I was not...). Just pulled the correct files and works awesomely well. No issues found.

Just tested with a small dictonary (like 100 words) -- works as expected... so far.

PS We should increment the version (in each commit so we know that we have the correct branch/version).

**Comment by JuliusSweetland** [comment]
That's excellent, thanks Andy. In terms of versioning; I'm vaguely using
semantic versioning, so no part of the version number is incremented with
commits. It would also lead to a range of disconnected version numbers
across different feature branches if I did.

As part of the usual git workflows you should sync (pull/fetch) before you
start do you always have/are aware of where you are relative to the head of
the branch.

**Comment by JuliusSweetland** [comment]
@andylamp I haven't individually evaluated any of theses files yet, but these pages contains some potentially useful dictionary files:

http://www.translatum.gr/dictionary/download-greek.htm
http://www.worldlingo.co.uk/en/resources/greek_dictionaries.html

Any good?

**Comment by andylamp** [comment]
@JuliusSweetland I've checked the links but on the first one I could not download the dictionary as a `.dic` file, it only had words -- seemed fine for most commonly used words though.

The second one is similar to the first, adequate.

If they are OK license-wise then we can use either of them.

**Comment by JuliusSweetland** [comment]
Great stuff. How many entries are there?

**Comment by andylamp** [comment]
@JuliusSweetland they have multiple dictionaries and I've only checked those that are 4k-5k words approximately (it's stated by them). They should work fine -- size wise.

**Comment by JuliusSweetland** [comment]
Any concerns that they're too small? English is 75k and Dutch goes up to
300k, which is a bit too big, but gives you an idea of target size.

**Comment by andylamp** [comment]
@JuliusSweetland I have no idea on this stuff; for sure it will be small. I mean even the first one that I've checked (the one that had 750k entries) was inadequate as Greek has a lot of words that mean the same thing that are used frequently. Tailoring a good `.dic` is not an easy task and very labor intensive. We can try to search dictionaries around the 100k mark (unsure about how to judge the quality of them) but we can initially deploy a smaller one.

**Comment by JuliusSweetland** [comment]
Ok. I'll take a look early next week. I'm nearly done customising the other
language keyboard so as soon as we agree a dictionary I think we're good to
start testing for release.

**Comment by andylamp** [comment]
@JuliusSweetland Super stuff! Let me know if there is anything else that I can help with -- coding wise.

**Comment by JuliusSweetland** [comment]
Will do, thanks. It would be really helpful if you got the latest from the
Greek branch and took a look at the updated language keyboards. I haven't
don't German and Russian yet, but the others are there. I've added dead key
diacritics and changed the layouts of the alpha and conversation alpha
keyboards a bit to accommodate the new keys. Let me know what you think.

**Comment by andylamp** [comment]
@JuliusSweetland Ok -- do that changes you want in the Greek branch; I'll pull the latest this afternoon and I'll take a look. I think we have to elaborate a bit on how to handle polytonic languages as if we want to support Arabic, Chinese and so on this "hack-n-slash" method of hard-coding everything will be a pain (and a huge source of potential bugs).

Let me know what you think.

**Comment by JuliusSweetland** [comment]
@andylamp Have you found a suitable Greek dictionary that we can use? Everything else on the Greek branch is nearly done and ready for beta testing. I've had a look for Greek dictionaries, but I don't read Greek so it's impossible for me to assess the quality of what I find.

**Comment by andylamp** [comment]
@JuliusSweetland can you link the ones you found so that I can test them (the .dic files). The ones that I found are really bad...

**Comment by JuliusSweetland** [comment]
The only ones that I found should have been posted above on this thread. I
can't check right now.

**Comment by JuliusSweetland** [comment]
@andylamp I've just released v2.3.1_pre-release_1 (https://github.com/OptiKey/OptiKey/releases) which isn't quite finished (I'm waiting for a few strings to be translated), but has a 100k Greek word dictionary which was generated from the 100k most common words used in Greek tv shows and films. Can you try it out and let me know what you think? N.B. If you've already run OptiKey with a test Greek dictionary then you will have to manually delete your customised/user Greek dictionary - see here: https://github.com/OptiKey/OptiKey/wiki/Factory-reset-the-dictionary

**Comment by andylamp** [comment]
@JuliusSweetland Okay will test today and see what's what... (remember that I crafted a build command inside the solution of OptiKey specifically for cleaning user files :))

**Comment by JuliusSweetland** [comment]
@andylamp Hi Andy - I've just noticed the two closed PRs (https://github.com/OptiKey/OptiKey/pull/239 and https://github.com/OptiKey/OptiKey/pull/240) - I presume these might have included the new string translation? Could you post it here and I'll add it manually? Thanks.

**Comment by andylamp** [comment]
@JuliusSweetland better get the updated resx from Transifex as I've corrected a couple of more strings to be more concise. During testing the -pre build I found the following two issues:

* a "cancel" windows sound pops right after a letter that can be enhanced by punctuation is pressed. (no idea why)

* the syntax engine seems to prefer words that have no punctuation than those who do. For example I type `ωραι` I get prompted to complete with `ωραιο` instead of the correct version which is `ωραίο`

**Comment by JuliusSweetland** [comment]
@andylamp The "cance" sound is probably because OptiKey has focus and when you select a key it is trying to type onto itself. As OptiKey cannot receive typed text it sounds an error tone. Try pressing physical keys with OptiKey focused and you should hear the same tones?

Regarding the suggested words; could you check the GreekGreece.dic file and look for instances of `ωραιο` and/or `ωραίο` - I'm thinking the correct version may not be in there?

**Comment by andylamp** [comment]
@JuliusSweetland 

* Uhm, basically after a restart clean/rebuild didn't do those sounds. I'll look into it a bit more.

* it's there but it's not suggested.

Also I noticed that there is no prediction at all based on the context of the conversation, like in SwiftKey for example. Can we implement something there? Intel's one has that functionality, maybe we could implement something along those lines?

**Comment by JuliusSweetland** [comment]
@andylamp contextual word suggestion is out of scope for this specific
change. There is an issue to integrate presage, but it's a different piece
of work. What about the diacritic/non diacritic suggestion issue from
yesterday?

**Comment by andylamp** [comment]
@JuliusSweetland still present. not suggesting the correct keywords with punctuation. Just try and type the example I gave you -- there is no need to know Greek just type the word using the characters you see, one at a time and see the suggestions.

**Comment by JuliusSweetland** [comment]
@andylamp ok will do. Is the version without the diacritic marks completely
incorrect, i.e. is it never a word?

**Comment by andylamp** [comment]
@JuliusSweetland just try the example I gave you, you'll immediately understand what I am saying. Specifically when the word builds up as you type letter you'll get to a point where `ωραι` and you won't be suggested `ωραίο` but `ωραί` or `ωραιο` which is the correct word w/o punctuation. I hope I get through the message as it's quite hard to relay that if you don't know Greek :).

**Comment by JuliusSweetland** [comment]
I understand. I'll try it asap.

**Comment by andylamp** [comment]
@JuliusSweetland also when you have the time, give me your thoughts on what I said about predictive text -- which is to provide a SwiftKey-like functionality on the keyboards.

Additionally the dictionary sizes are going to pose a huge issue with rich languages like Japanese, Chinese and so on... entry count will be much, much higher; so I think something needs to be reworked  there. I can help, but can you provide a short run-down on how dictionaries are parsed and handled so that I don't have to dig all the logic straight from the code... if at all possible -- but we should create a separate issue for that.

Finally, since what essentially is left to do here are tweaks with the dictionary shall I start working on retouching the translation how-to docs?

**Comment by JuliusSweetland** [comment]
@andylamp Here are the results of my testing with a clean dictionary and build (of the Greek branch):

* Typing ωρα results in these 4 suggestions:
![image](https://cloud.githubusercontent.com/assets/4722561/14153019/0b13bcd4-f6ae-11e5-9adb-d88e95ce7ea3.png)

* BUT, the next suggestions are:
![image](https://cloud.githubusercontent.com/assets/4722561/14153032/1a9d4aa8-f6ae-11e5-8c69-bf21e96ea3e8.png)

* If I continue with ι then the suggestions are:
![image](https://cloud.githubusercontent.com/assets/4722561/14153049/2b39bafe-f6ae-11e5-99a4-d1855cbf4935.png)

* The next suggestions are:
![image](https://cloud.githubusercontent.com/assets/4722561/14153059/37534922-f6ae-11e5-9225-02522f09ea3d.png)

* If I remove the ι and instead type ί then the suggestions are:
![image](https://cloud.githubusercontent.com/assets/4722561/14153084/4f09f9ee-f6ae-11e5-9c08-ebf2036b63a0.png)

I think this looks about right. The only issue may be that ωραιο isn't a valid word and should be removed from the dictionary? Let me know.

RE word prediction: This issue already exists regarding next word prediction using Presage - https://github.com/OptiKey/OptiKey/issues/69

RE symbolic/polytonic languages: I have NO idea how large the problem will be for these languages. I suspect there will need to be many changes. For the moment I'm going to keep the focus on latin languages (and other character sets that work in the same way). There are lots of languages that can easily be added before lots of time is spent on Chinese/Japanese/etc. If, however, you'd like to start on this problem anyway then I can create an issue and give you some guidance. Let me know.

**Comment by andylamp** [comment]
@JuliusSweetland 

1) Well, we have to refine this as I am annoyed if the first couple words suggested do not include the one I want -- let alone someone with a disability... I think by the time you've typed `ωρα` you should've *at least* got `ώρα` and `ωραίο` in the first three matches as both `ώρα` and `ωραίο` are very commonly used words. Additionally, you'll probably have to take my word on this one as I do speak the language and it's really hard to convey the message of "how-good" the prediction is without knowing that language to some extent (that applies for me as well for other languages that I don't speak! ;))

2) Okay, I'll look into that issue and I'll pitch in afterwards.

3) Yes, please do.

**Comment by JuliusSweetland** [comment]
@andylamp Understood, but I don't understand what is wrong with these suggestions:
![image](https://cloud.githubusercontent.com/assets/4722561/14153019/0b13bcd4-f6ae-11e5-9adb-d88e95ce7ea3.png)
The first suggestion is always what the user has typed. This is because I have seen users who only use the suggestion to select what they have typed and leverage the auto-space functionality. I don't think this is the best way to use OptiKey, but I have to cater for real world use cases. The other suggestions are ordered by a few things, but mainly shorter words to longer words. 

RE symbolic languages - we can discuss this on a new issue if you want to create one, but I'm going to focus on getting Greek out first so I may be a bit distracted in the short term.

**Comment by andylamp** [comment]
@JuliusSweetland I'll try to elaborate. The biggest issue is that for example -- and I *really* using *simple* prediction cases -- `ωρα` is an actually word; so immediately in the first results you should have what you typed and `ώρα`, which is the same word but using punctuation (w/o punctuation you have what you typed already, which is `ωρα`). Then I would expect to have `ωραίο` which is the word we want but with correct punctuation -- it's unreasonable to think that one would prefer a word w/o the correct punctuation to be in higher priority than the one that is. 

Does that make sense? :)

**Comment by JuliusSweetland** [comment]
@andylamp it does make sense, but the suggestions are not contextual.
OptiKey currently can only look at the current word that you're typing and
make suggestions on that in isolation.

**Comment by andylamp** [comment]
@JuliusSweetland can that be enhanced so that it does? Otherwise, if that's the case I don't think we can do anything more; other than prefer punctuated and complete words instead of half-baked or w/o punctuation. If these can be fixed you can release it... if not then, release it as is and we make refinements later because I get the feeling we have to work on this for quite a while.

**Comment by JuliusSweetland** [comment]
@andylamp Word prediction CAN be add (see the Presage issue), but it's a
separate piece of work to the Greek release.

In terms of which words are suggested: are any of the words that are
currently suggested incorrect words, i.e. are any of the suggestions not
valid words? If so then we need to clean up the dictionary manually.

**Comment by andylamp** [comment]
@JuliusSweetland As I said, the best thing to do at the moment is to hide half-baked (e.g. `ωραι` instead of `ωραιο`) and non-punctuated words entirely.

**Comment by JuliusSweetland** [comment]
@andylamp If ωρα, ώρα, ωραι, ωραί, ωραίο and ωραιο are all valid words in one scenario or another, then there is currently no way for OptiKey to know this and all might be suggested. If any of them are NEVER valid words then they need to be removed from the GreekGreece.dic file completely.

**Comment by andylamp** [comment]
@JuliusSweetland `ώρα` and `ωραίο` are valid words -- the other two are not. Trimming the dictionary is a considerable task and a very tedious process. I'd avoid that if possible.

**Comment by JuliusSweetland** [comment]
@andylamp I'll write something which extracts all entries which have multiple entries which only differ by their accents. Perhaps the list won't be as large as we fear. Otherwise we need a new dictionary.

**Comment by JuliusSweetland** [comment]
@andylamp Also I still need to translate this string - can you please correct:

Greek: "Καταστολή πλήκτρα τροποποίησης για όλες τις δράσεις του ποντικιού:"

**Comment by andylamp** [comment]
@JuliusSweetland isn't that on transifex? Pull the resx file from there as I think I translated that already.

**Comment by JuliusSweetland** [comment]
@andylamp ah ok. Will do, thanks.

**Comment by andylamp** [comment]
@JuliusSweetland yeah, because I've changed a couple of other strings as well -- grab that from there just to be sure it's up to date.

**Comment by andylamp** [comment]
@JuliusSweetland you didn't tell me if I should start reworking the translation how-to... If anything else is needed to be done in this issue before we release let me know, so I can do it -- otherwise I am going to start ;).

**Comment by JuliusSweetland** [comment]
@andylamp I think you're good to go. For this Greek release the only
outstanding tasks are: to clean up the dictionary, add the new setting
string for French and Dutch, and test thoroughly for all languages as the
dictionaries/keyboard layouts/method of typing diacritics has changed.

**Comment by JuliusSweetland** [comment]
@andylamp Ok, I've knocked up a LinqPad program which looks for duplicate entries in a dictionary file. Duplicates are entries which would be the same if the diacritics (accents etc) were removed. The 100k Greek dictionary has 13,284 groups of duplicates, so the problem is significant. Here is the linqpd file - copy this into a text editor and save it as SomeFileName.linq. You can then open this in LinqPad (download it if you don't have it - it's free and it's useful). You'll have to change the 'inputFile' variable to point to your own copy of the GreekGreece.dic file, then run it. It will dump out the first 1000 duplicated entries. Let me know what you think we should do as I am not sure which are valid duplicates, and which are completely incorrect entries which would need to be removed.
![image](https://cloud.githubusercontent.com/assets/4722561/14189470/5770e0de-f785-11e5-8a7d-f4693447b1eb.png)

**Comment by JuliusSweetland** [comment]
@andylamp One thought would be to simply discard any duplicated strings which do not have diacritics. This may be overly simplistic though and result in entries that shouldn't have diacritics being kept over the more correct form, but I need your input on that. If successful we should have far fewer duplicates to look at (where there are more than 2 entries with different diacritics).

**Comment by JuliusSweetland** [comment]
@andylamp I've just checked a bit deeper:

10,140 duplicate sets where one entry has no diacritics, while the other entry has at least one diacritic. These could be reduced by favouring the version with diacritics (if you think this is a valid way forward).

1,796 duplicate sets with more than 2 duplicate entries. These can't be automatically cleaned so these require manual inspection.

**Comment by andylamp** [comment]
@JuliusSweetland I can try to clean this up, but I can't really do it that fast... will take a week as it's cumbersome and tiring task. What I propose is to have only words that are relevant to *modern-Greek* and are punctuated. So for example `είναι` would be OK and all others would be rejected -- but a Greek has to do that so you know semantically what to discard and what not...

**Comment by JuliusSweetland** [comment]
@andylamp would you like me to provide an initially cleaned version where
duplicated entries without diacritics are removed?

**Comment by andylamp** [comment]
@JuliusSweetland yep, I can do that -- the question is I don't know how fast I can do that. I'll try my best.

**Comment by JuliusSweetland** [comment]
@andylamp let me know if I can help. Shall we catch up next week to see how
you're going? In the mean time I'll ask other contributors to test the
other languages.

**Comment by andylamp** [comment]
@JuliusSweetland yea for sure. I wish you could help, but as you might imagine you have to know the language to understand what to discard and what not :/.

**Comment by andylamp** [comment]
@JuliusSweetland I am progressing slowly but steadily. Hopefully I'll be done by weeks end or close to that.

PS It's **very** monotonous

**Comment by JuliusSweetland** [comment]
@andylamp that's good to hear. I totally understand the monotony. I've done
a few long winded tasks like this!

**Comment by JuliusSweetland** [comment]
@andylamp Hi Andy - how is the cleanup going? Still fun?!

**Comment by andylamp** [comment]
@JuliusSweetland for sure, it's quite a lot I am doing like 80 entries a day... that's all the time I can spare :).

**Comment by JuliusSweetland** [comment]
@andylamp Understood. It's long winded. Do you know how far through you are
now?

**Comment by andylamp** [comment]
@JuliusSweetland like... under a thousand for sure and there are more than 10k.

**Comment by JuliusSweetland** [comment]
*ALL* Can anyone help @andylamp clean the Greek dictionary? As soon as this last task is done we can release Greek and Turkish, and move on to Spanish.

**Comment by JuliusSweetland** [comment]
@andylamp Ok, I've tried something new with the Greek dictionary. I downloaded a Hunspell Greek dictionary file, which is used in Firefox/OpenOffice/etc, extracted the .dic and .aff files, unmunched them into a flat file, and then cleaned & extracted the top 100,000 entries, ordered by frequency of use in tv show and movie scripts. This essentially is the same as using the dictionary we already have, but I only took entries which appear in the official Hunspell Greek dictionary, so we can be confident that they are valid entries.

The new file has been pushed to my Greek branch and there are now 2,529 duplicates (with 2 or more duplicates differing only by their accents). Of these 57 have 3 duplicate entries, and 2 have 4 duplicate entries.

I tried googling a few of these duplicates, but couldn't tell if they were legit or not.

Can you let me know what you think? I'm really hopeful that this method might produce a strong dictionary of clean entries which can be used for future languages.

**Comment by andylamp** [comment]
@JuliusSweetland sure, let me check it out and I'll tell you my thoughts on it. For sure cleaning top 2k entries is *much* easier than having to clean 13k.

**Comment by JuliusSweetland** [comment]
@andylamp What do you think of the new Greek dictionary? Does it look better?

**Comment by andylamp** [comment]
@JuliusSweetland sorry I was swamped the last few days and I haven't have the chance to check it out. I'll take a look today and probably will let you know today or tomorrow.

**Comment by JuliusSweetland** [comment]
@andylamp great, thanks

**Comment by andylamp** [comment]
@JuliusSweetland I've tested the dictionary (in it's pulled state) and it's really *much* better than the previous one -- go ahead and release it I don't we can do much better just on the dictionary side without doing some semantic analysis of the context. Also I'd like to add a feature -- as an option (turned one by default?). Once the user presses to complete a matching word we should add a space after the matched word unless the next character is a punctuation point (e.g. comma, dot etc). This saves at least one space key-press per word and it's not that difficult to implement.

**Comment by JuliusSweetland** [comment]
Thanks @andylamp I'll try to get a release out this weekend if I can.

Regarding auto-space; this feature is already in OptiKey. If you use an auto-complete or multi-key selection suggestion to type a word, then you can continue typing manually, or perform another multi-key selection and a space will be inserted automatically for you between the word. Give it a try.

**Comment by andylamp** [comment]
@JuliusSweetland oh, it's there -- I was used with the paradigm that when you auto complete the word the space is added then and removed if needed (like in most auto-complete keyboards). So after I was completing the word I just pressed space myself instead of just continuing to type.

As a suggestion, because I find it counter intuitive this way; what do you think of adding the space proactively after auto-completing the word and not after you press another letter? We can always remove it if needed.

**Comment by JuliusSweetland** [comment]
@andylamp both ways are possible. I went for the current approach as that's
how the Swype keyboard on phones works and I like it. I make an effort in
the getting started guide to point it out to new users so they are aware of
it.

If your wanted to implement the other approach I'd be happy to merge in the
change, but I'm reluctant to make this change myself as the logic settings
auto spacing alongside suggestions, manual typing, back spaces, etc is
complex and, dare I say, fragile.

**Comment by andylamp** [comment]
@JuliusSweetland that's nice! Well, we can make it as an option to the user to select either `Swype`-type space auto-completion or `SwiftKey`-type. But this will have to wait a bit as I want to finish the translation document how-to during the weekend so that others can perform the translations a bit easier.

**Comment by andylamp** [comment]
@JuliusSweetland I am about to finish the new translation guide, since it's only one file I don't think we need to create a new branch for it... so instead I think I'll just push it in the `Greek` branch for you to review it. Are you OK with this?

PS likely it will be done tomorrow.

**Comment by JuliusSweetland** [comment]
Releasing Greek (shortly) as v2.3.1

**Comment by JuliusSweetland** [comment]
@andylamp Please can you make and PR the localisation document change against master when you are ready. I have merged and deleted the Greek branch. Thanks.

**Comment by andylamp** [comment]
@JuliusSweetland will do.
